src/System/query_failures.rs

```diff
@@ -0,0 +1,645 @@
+                f,
+                "The refinement '{}' fails because '{}' is empty",
+                system.name, child
+            ),
+            RefinementPrecondition::EmptyInitialState { system } => write!(
+                f,
+                "The refinement '{}' fails because it has no initial state",
+                system.name
+            ),
+            RefinementPrecondition::InconsistentChild(cf, system) => write!(
+                f,
+                "The refinement '{}' fails because of inconsistent system: {}",
+                system.name, cf
+            ),
+            RefinementPrecondition::ActionMismatch(action, system) => {
+                write!(
+                    f,
+                    "The refinement '{}' fails based on actions: {}",
+                    system.name, action
+                )
+            }
+        }
+    }
+}
+
+// ------------------------------- //
+// - Ugly conversions begin here - //
+// ----- You have been warned ---- //
+// ------------------------------- //
+mod conversions {
+    use super::*;
+    use std::error::Error;
+    impl Error for SystemRecipeFailure {}
+    impl Error for ClockReductionFailure {}
+    impl Error for RefinementFailure {}
+    impl Error for ConsistencyFailure {}
+    impl Error for DeterminismFailure {}
+
+    impl From<RefinementPrecondition> for RefinementFailure {
+        fn from(failure: RefinementPrecondition) -> Self {
+            RefinementFailure::Precondition(failure)
+        }
+    }
+
+    impl From<CompositionType> for Composition {
+        fn from(comp: CompositionType) -> Self {
+            match comp {
+                CompositionType::Quotient => Composition::Quotient,
+                CompositionType::Composition => Composition::Composition,
+                CompositionType::Conjunction => Composition::Conjunction,
+                CompositionType::Simple => Composition::Simple,
+            }
+        }
+    }
+
+    impl From<DeterminismFailure> for ConsistencyFailure {
+        fn from(failure: DeterminismFailure) -> Self {
+            ConsistencyFailure::NotDeterministic(failure)
+        }
+    }
+
+    impl From<SystemRecipeFailure> for QueryResult {
+        fn from(res: SystemRecipeFailure) -> Self {
+            QueryResult::RecipeFailure(res)
+        }
+    }
+
+    impl From<PathResult> for QueryResult {
+        fn from(res: PathResult) -> Self {
+            QueryResult::Reachability(res)
+        }
+    }
+
+    impl From<ConsistencyResult> for QueryResult {
+        fn from(res: ConsistencyResult) -> Self {
+            QueryResult::Consistency(res)
+        }
+    }
+
+    impl From<DeterminismResult> for QueryResult {
+        fn from(res: DeterminismResult) -> Self {
+            QueryResult::Determinism(res)
+        }
+    }
+
+    impl From<RefinementResult> for QueryResult {
+        fn from(res: RefinementResult) -> Self {
+            QueryResult::Refinement(res)
+        }
+    }
```

Comment:
A lot of duplicate(ish) code here. If we implement the macro specified previously, maybe do the same here?